### PR TITLE
NavBar and React Routing brought together. Styling tackled.

### PR DESCRIPTION
**Pull request by mwilkins74**
Navigation handled by bringing NavBar/NavList and React Routing together. 
Tackled styling too!
Clicking on a Link highlights the Link in the (currently) light grey div.

Screen Shot 2024-01-09 at 5 25 32 PM

### CODE REVIEW

File: src/components/molecules/NavList.tsx [hunk 1]
```diff
@@ -1,53 +1,187 @@
-import React from 'react';
-import {List, ListItemButton, ListItemText, ListSubheader} from '@mui/material';
+import React, { useState } from "react";
+import {
```
Comment (kmdunn5) [inline src/components/molecules/NavList.tsx:2]: Can we compress this? I think the imports on one line is fine. I think VSCode dulls the imports that aren't being used anyway (unless thats a plugin I have) so it'll be easy to tell.

File: src/components/molecules/NavList.tsx [hunk 2]
```diff
@@ -1,53 +1,187 @@
...
+      borderRadius: "0 24px 24px 0",
+      "&:hover": {
+        backgroundColor: "lightgrey",
+      },
+    }),
+  };
+}
+
+function NavList() {
+  const [selectedLink, setSelectedLink] = useState<string | null>(null);
```
Comment (kmdunn5) [inline src/components/molecules/NavList.tsx:31]: As discussed, we'll get this from `useLocation` or our `useQuery`

File: src/components/molecules/NavList.tsx [hunk 3]
```diff
@@ -1,53 +1,187 @@
...
+          primary="Video List"
+          sx={{
+            color: isLinkSelected("/video-list-page") ? "black" : "white",
+            padding: "1px",
+          }}
+        />
       </ListItemButton>
-      
+
       {/* WIP style sub items */}
```
Comment (kmdunn5) [inline src/components/molecules/NavList.tsx:101]: Is this still in progress? if not we can remove this comment

File: src/components/molecules/NavList.tsx [hunk 4]
```diff
@@ -1,53 +1,187 @@
...
+import { Link } from "react-router-dom";
+
+const commonStyle = {
+  fontWeight: "bold",
+  bgcolor: "black",
+  color: "white",
+  fontSize: 20,
+};
+
+function generateButtonStyle(isSelected: boolean) {
```
Comment (kmdunn5) [inline src/components/molecules/NavList.tsx:17]: I like this

File: src/components/molecules/NavList.tsx [hunk 5]
```diff
@@ -1,53 +1,187 @@
...
       </ListItemButton>
-      <ListItemButton>
-        <ListItemText primary="Contact" />
+      <ListItemButton
+        component={Link}
+        to="/video-list-page"
+        sx={{ ...generateButtonStyle(isLinkSelected("/video-list-page")) }}
+        onClick={() => handleLinkClick("/video-list-page")}
+      >
+        <ListItemText
```
Comment (kmdunn5) [inline src/components/molecules/NavList.tsx:92]: Won't be an actual nav item, but the video list will be the actual rendered component for most other routes

File: src/components/molecules/NavList.tsx [hunk 6]
```diff
@@ -1,53 +1,187 @@
...
+}
+
+function NavList() {
+  const [selectedLink, setSelectedLink] = useState<string | null>(null);
+
+  const handleLinkClick = (link: string) => {
+    setSelectedLink(link);
+  };
+
+  const isLinkSelected = (link: string) => selectedLink === link;
```
Comment (kmdunn5) [inline src/components/molecules/NavList.tsx:37]: Can the color that this typically sets be included in the `generateButtonStyle` function?

File: src/components/molecules/NavList.tsx [hunk 7]
```diff
@@ -1,53 +1,187 @@
...
       </List>
     </List>
-
-  )
+  );
 }
 
-export default NavList
\ No newline at end of file
+export default NavList;
```
Comment (kmdunn5) [inline src/components/molecules/NavList.tsx:187]: We can move this export to above

**Review by kmdunn5 — CHANGES_REQUESTED** [review]

File: src/components/molecules/NavList.tsx (hunk 3 above)
Comment (mwilkins74) [reply src/components/molecules/NavList.tsx:101]: I did touch styling in this PR. 
Mostly just the coloration of the NavBar so this has ben _sorta_ progressed.

File: src/components/molecules/NavList.tsx (hunk 6 above)
Comment (mwilkins74) [reply src/components/molecules/NavList.tsx:37]: No idea, but possibly!

File: src/components/molecules/NavList.tsx (hunk 7 above)
Comment (mwilkins74) [reply src/components/molecules/NavList.tsx:187]: Yup. I missed this one.

File: src/components/molecules/NavList.tsx (hunk 1 above)
Comment (mwilkins74) [reply src/components/molecules/NavList.tsx:2]: Compressed and removed the one "dull" import.

File: src/components/molecules/NavList.tsx (hunk 6 above)
Comment (mwilkins74) [reply src/components/molecules/NavList.tsx:37]: The answer is yes!

File: src/components/molecules/NavList.tsx [hunk 8]
```diff
@@ -1,53 +1,107 @@
...
-        <ListSubheader component="div">
-          ROLES
+      <List component="div">
+        <ListSubheader component="div" sx={{ ...commonStyle }}>
+          ROLE
         </ListSubheader>
-        <ListItemButton>
+        <ListItemButton
+          component={Link}
+          to="/director-producer"
```
Comment (kmdunn5) [inline src/components/molecules/NavList.tsx:65]: I think @ChuckBTaylor made a comment about these routes in another PR. We're going to put a filter on them.

**Review by kmdunn5 — DISMISSED** [review]

File: src/components/molecules/NavList.tsx (hunk 8 above)
Comment (mwilkins74) [reply src/components/molecules/NavList.tsx:65]: Yes!
I think I overlooked since it was talked about on another PR. 
I'll address later today.

Comment (mwilkins74) [reply src/components/molecules/NavList.tsx]: Ok, changes made to add filter. 
Now reads like this: `/video-list?filter=director-producer`

Clicking on one of the NavList items returns this URL:
`http://localhost:3000/video-list?filter=documentary`

File: src/components/molecules/NavList.tsx
Comment (ChuckBTaylor) [inline src/components/molecules/NavList.tsx:1]: 🔧 This component feels wonky to me.  Not sure if it's the repetitive css, the Links, or what.

I think we can break this down further.  To me, this while thing is an organism. Each button is an atom. The button group (w/ header) is the molecule, and then all together we have an organism

Comment (mwilkins74) [reply src/components/molecules/NavList.tsx:1]: I can begin deconstructing and moving things into smaller pieces.

**Review by ChuckBTaylor — APPROVED** [review]

**Comment by mwilkins74** [comment]
Still working on that last commit (aka draft). Wanted to commit it my current work just in case I don't get it completed today and don't touch till next week.

**Comment by mwilkins74** [comment]
Ok, Navigation Buttons have been all pulled out of NavList and restructured. 
It is super messy as they are all in their own files right now as atoms. 
Will absolutely help to utilize the enum stuff here to clean all this up. 
That will be a next week task for me at least.

File: src/common/utils/reactUtils 2.ts [hunk 9]
```diff
@@ -0,0 +1,7 @@
+import { useMemo } from "react";
+import { useLocation } from "react-router";
+
+export function useQuery() {
+  const { search } = useLocation();
+  return useMemo(() => new URLSearchParams(search), [search]);
+}
```
Comment (ChuckBTaylor) [inline src/common/utils/reactUtils 2.ts:7]: 🔧 We can use the react-router-dom provided function `useSearchParams` to to get this result

File: src/components/atoms/cameraOperatorButton/CameraOperatorButton.tsx
Comment (ChuckBTaylor) [inline src/components/atoms/cameraOperatorButton/CameraOperatorButton.tsx:1]: 🔧 I think that `CameraOpreatorButton` is too specific. We can have a single button atom that works for each of the buttons in the nav bar

File: src/components/molecules/genreListSubheader/GenreListSubheader.tsx
Comment (ChuckBTaylor) [inline src/components/molecules/genreListSubheader/GenreListSubheader.tsx:1]: 🔧 Same idea here. We don't need a `GenreListSubheader` and a `RoleListSubheader`. Just one that displays "Genre" or "Role"

Comment (mwilkins74) [reply src/components/molecules/genreListSubheader/GenreListSubheader.tsx:1]: Agreed. 
As I mentioned in my comment this work was to just get something up. 
Clean up, reorganization, concise code will happen later.

**Comment by mwilkins74** [comment]
Pull apart `NavList` and made it more concise. 
I made a `NavLinks` components to contain Recent Work, About and Contact and a `SubheaderList` component for Roles and Genres. (Now that I'm typing them the names of these components need changing).
I couldn't figure out how to do one component for all of the above Links, but this is much cleaner so wanted to send it up for starting to piece the app together and moving forward. 

Styling will also need fixing but we all know that.

File: src/components/molecules/NavList.tsx [hunk 10]
```diff
@@ -1,53 +1,91 @@
...
+
+  const generateButtonStyle = (isSelected: boolean) => ({
+    backgroundColor: isSelected ? 'lightgrey' : 'transparent',
+    padding: '1px 16px',
+    borderRadius: isSelected ? '24px 0 0 24px' : 'unset',
+    '&:hover': {
+      backgroundColor: 'lightgrey',
+    },
+    color: isSelected ? 'black' : 'white',
+  })
```
Comment (ChuckBTaylor) [inline src/components/molecules/NavList.tsx:57]: ❓ Why have this be a function defined within the component that has to be passed down? Can we exrtact it to some cssHelper file?

**Review by ChuckBTaylor — COMMENTED** [review]
Looks to me like your treating each section of the nav menu as it's own molecule, which I think makes sense.

And reason to treat the top nav items different from the ones that have a header?

I'd imagine that the (sub)header would be an atom and each link would be an atom.  Liiks like you're using the MUI `ListItemButton` as the item which I think makes sense.

But why have two different molecules for the different sections? Could we create one molecule that conditionally renders a subheader if it's given?

**Review by kmdunn5 — APPROVED** [review]
Happy to get this merged in and do some code tweaks on it in a different PR.

**Comment by mwilkins74** [comment]
Sure. That was my intention. Get it better than it was, merge it in, and keep moving forward.

**Comment by mwilkins74** [comment]
I will merge this in now. 
Then would love to see/some guidance on how to make this even more intelligently written.

**Pull request merged**